File: backend-express/src/models/cruds/place.ts
import { EntityManager } from "typeorm";

import { placeEmbedding } from "@/background/publishers";
import { env } from "@/config";
import { CrudsClass } from "@/lib/typeorm_";
import { ApiError, HttpStatus } from "@/lib/types";
import { huggingFace, pgClient, storage } from "@/services/instances";

import { Models, Place } from "../orm";
import {
    PlaceCreate,
    PlacePost,
    PlacePut,
    PlaceRead,
    PlaceSearchableType,
    PlaceSearchQuery,
    placeSelectableFields,
    PlaceSelectableType,
    PlaceSortableType,
    PlaceUpdate,
    UserRead,
} from "../schemas";
import { userExists } from "./utils";

export type PlaceOptions = {
    process?: boolean;
    fields?: PlaceSelectableType[];
};

export class CrudsPlace extends CrudsClass<
    Place,
    UserRead,
    PlaceCreate,
    PlacePost,
    PlaceRead,
    PlaceSelectableType,
    PlaceSortableType,
    PlaceSearchableType,
    PlaceUpdate,
    PlacePut,
    PlaceOptions
> {
    MAX_ITEMS_PER_PAGE = env.MAX_ITEMS_PER_PAGE;

    // Post-Processing

    async postProcess<T extends Partial<PlaceRead> | PlaceRead>(
        raw: T
    ): Promise<T> {
        if (raw.imageUrl) {
            raw.imageUrl = await storage.getSignedUrl(raw.imageUrl);
        }
        return raw;
    }

    // Query Building

    mapWhere(field: string): string {
        switch (field) {
            case "locationLat":
                return `(location->>'lat')::float`;
            case "locationLng":
                return `(location->>'lng')::float`;
            default:
                return this.fieldAlias(field);
        }
    }

    // Create

    async afterCreate(
        _manager: EntityManager,
        id: number,
        _data: PlaceCreate,
        _context: Record<string, any>
    ): Promise<void> {
        placeEmbedding(id);
    }

    async seed(data: PlaceCreate, embedding: number[]): Promise<number> {
        // Used when seeding the dev/test database
        // Avoid triggering the place embedding
        const id = await super.create(data);
        await this.updateEmbedding(id, embedding);
        return id;
    }

    async postToCreate(form: PlacePost): Promise<PlaceCreate> {
        const imageUrl = await storage.uploadFile(form.image || null);
        const { image: _image, lat, lng, ...body } = form;
        return { ...body, imageUrl, location: { lat, lng } };
    }

    async authPost(user: UserRead, data: PlacePost): Promise<void> {
        if (!user) {
            throw new ApiError(HttpStatus.UNAUTHORIZED, "Not Authenticated");
        }

        if (user.isAdmin) {
            const exists = await userExists(this.datasource, data.creatorId);
            if (!exists) {
                throw new ApiError(HttpStatus.NOT_FOUND, "User not found", {
                    message: `Cannot set creatorId to ${data.creatorId}, No user with id ${data.creatorId} found in the database`,
                });
            }
        }

        if (user.id != data.creatorId) {
            throw new ApiError(HttpStatus.UNAUTHORIZED, "Access denied", {
                message: `Cannot add places to user ${data.creatorId}`,
            });
        }
    }

    // Read

    authGet(user: UserRead, query: PlaceSearchQuery): PlaceSearchQuery {
        if (!query.where) query.where = {};
        query.where.creatorId = this.eq(user.id);
        return query;
    }

    // Update

    async beforeUpdate(
        manager: EntityManager,
        id: number,
        data: PlaceUpdate
    ): Promise<Record<string, any>> {
        const record = await manager.findOne(Place, {
            where: { id },
            select: ["title", "description"],
        });
        if (!record) {
            throw this.notFoundError(id);
        }

        const titleChanged = !!data.title && data.title !== record.title;
        const descriptionChanged =
            !!data.description && data.description !== record.description;
        if (descriptionChanged || titleChanged) {
            return { triggerEmbedding: true };
        }
        return { triggerEmbedding: false };
    }

    async afterUpdate(
        _manager: EntityManager,
        id: number,
        _data: PlaceUpdate,
        context: Record<string, any>
    ): Promise<void> {
        if (context.triggerEmbedding == undefined) {
            throw new ApiError(
                HttpStatus.INTERNAL_SERVER_ERROR,
                "something went wrong after trying to update place recod"
            );
        }

        if (context.triggerEmbedding) {
            placeEmbedding(id);
        }
    }

    async authPut(
        user: UserRead,
        id: number | string,
        _data: PlacePut
    ): Promise<void> {
        if (!user) {
            throw new ApiError(HttpStatus.UNAUTHORIZED, "Not Authenticated");
        }

        if (user.isAdmin) return;

        const check = await this.exists({
            id: this.eq(id),
            creatorId: this.eq(user.id),
        });
        if (!check) {
            throw new ApiError(HttpStatus.UNAUTHORIZED, "Access denied", {
                message: `Cannot not access place ${id}`,
            });
        }
    }

    async updateEmbedding(
        id: number | string,
        vector: number[]
    ): Promise<void> {
        await this.repository
            .createQueryBuilder()
            .update()
            .set({
                embedding: () => `'${JSON.stringify(vector)}'::vector`,
            })
            .where("id = :id", { id })
            .execute();
    }

    async embed(id: number): Promise<number[]> {
        const data = (await this.repository.findOne({
            where: { id },
            select: { title: true, description: true },
        })) as Pick<Place, "title" | "description">;
        if (!data) {
            throw new ApiError(
                HttpStatus.NOT_FOUND,
                `No place with id ${id} found in the database`
            );
        }
        const text = `${data.title} - ${data.description}`;
        const result = await huggingFace.embedText(text);
        try {
            await this.updateEmbedding(id, result);
        } catch (err) {
            if (err instanceof Error) {
                const status = HttpStatus.INTERNAL_SERVER_ERROR;
                const message = "embeddding failed";
                const details = { placeId: id, message: err.message };
                throw new ApiError(status, message, details);
            }
            throw err;
        }
        return result;
    }

    // Delete

    async afterDelete(_manager: EntityManager, obj: Place): Promise<void> {
        if (obj.imageUrl) {
            storage.deleteFile(obj.imageUrl);
        }
    }

    async authDelete(user: UserRead, id: number | string): Promise<void> {
        if (!user) {
            throw new ApiError(HttpStatus.UNAUTHORIZED, "Not Authenticated");
        }

        if (user.isAdmin) return;

        const check = await this.exists({
            id: this.eq(id),
            creatorId: this.eq(user.id),
        });
        if (!check) {
            throw new ApiError(HttpStatus.UNAUTHORIZED, "Access denied", {
                message: `Cannot not access place ${id}`,
            });
        }
    }
}

export const crudsPlace = new CrudsPlace(
    pgClient.client,
    Models.place,
    placeSelectableFields,
    ["createdAt"]
);

File: backend-fastapi/models/cruds/place.py
import json
import logging
from http import HTTPStatus
from typing import TypedDict, get_args

from pydantic import BaseModel
from sqlalchemy import Float, select, text, update
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import InstrumentedAttribute

from background.publishers import place_embedding
from lib.sqlalchemy_ import CrudsClass
from lib.types_ import ApiError, WhereFilters
from models.orm import Place
from models.schemas import (
    PlaceCreateSchema,
    PlacePostSchema,
    PlacePutSchema,
    PlaceReadSchema,
    PlaceSearchableFields,
    PlaceSearchQuery,
    PlaceSelectableFields,
    PlaceSortableFields,
    PlaceUpdateSchema,
    UserReadSchema,
)
from services.instances import cloud_storage, hf_client

from .utils import user_exists


class PlaceOptions(TypedDict):
    process: bool | None
    fields: list[PlaceSelectableFields] | None


class PlaceCreateContext(BaseModel):
    pass


class PlaceUpdateContext(BaseModel):
    trigger_embedding: bool


class PlaceDeleteContext(BaseModel):
    image_url: str


class CrudsPlace(
    CrudsClass[
        Place,
        UserReadSchema,
        PlaceCreateSchema,
        PlaceCreateContext,
        PlacePostSchema,
        PlaceReadSchema,
        PlaceOptions,
        PlaceSelectableFields,
        PlaceSortableFields,
        PlaceSearchableFields,
        PlaceUpdateSchema,
        PlaceUpdateContext,
        PlacePutSchema,
        PlaceDeleteContext,
    ]
):
    # Init

    def __init__(self, session: AsyncSession):
        super().__init__(
            session,
            Place,
            list(get_args(PlaceSelectableFields)),
            ["-createdAt"],
        )

    # Serialization and Post-Processing

    async def post_process(self, raw: PlaceReadSchema) -> PlaceReadSchema:
        if raw.imageUrl:
            raw.imageUrl = cloud_storage.get_signed_url(raw.imageUrl)
        return raw

    async def post_process_dict(self, raw: dict) -> dict:
        image_url = raw.get("imageUrl")
        if image_url:
            raw["imageUrl"] = cloud_storage.get_signed_url(image_url)
        return raw

    # Query Building

    def map_where(self, field: str) -> InstrumentedAttribute:
        if field == "locationLat":
            return self.model.location["lat"].astext.cast(Float)  # type: ignore
        elif field == "locationLng":
            return self.model.location["lng"].astext.cast(Float)  # type: ignore
        else:
            return super().map_where(field)

    # Create

    async def after_create(
        self, id: int, data: PlaceCreateSchema, context: PlaceCreateContext
    ) -> None:
        place_embedding(id)

    async def seed(
        self, data: PlaceCreateSchema, embedding: list[float]
    ) -> int:
        id = await self.create(data)
        await self.update_embedding(id, embedding)
        return id

    async def post_to_create(self, data: PlacePostSchema) -> PlaceCreateSchema:
        json = data.model_dump(exclude_none=True, exclude_unset=True)
        lat = json.pop("lat", None)
        lng = json.pop("lng", None)
        json["location"] = dict(lat=lat, lng=lng)

        image = json.pop("image", None)
        if image:
            json["imageUrl"] = cloud_storage.upload_file(image)
        else:
            json["imageUrl"] = ""

        return self.create_schema.model_validate(json)

    async def auth_post(
        self, user: UserReadSchema, form: PlacePostSchema
    ) -> None:
        if user.isAdmin:
            if not await user_exists(self.session, form.creatorId):
                raise ApiError(
                    HTTPStatus.NOT_FOUND,
                    "User not found",
                    dict(
                        message=f"Cannot set creatorId to {form.creatorId}, No user with id {form.creatorId} found in the database",
                    ),
                )
            return

        if user.id != form.creatorId:
            raise ApiError(
                HTTPStatus.UNAUTHORIZED,
                "Access denied",
                dict(message=f"Cannot add places to user {form.creatorId}"),
            )

    # Read

    def auth_get(
        self,
        user: UserReadSchema,
        query: PlaceSearchQuery,
    ) -> PlaceSearchQuery:
        if user.isAdmin:
            return query

        if query.where is None:
            query.where = {}

        query.where["creatorId"] = self.eq(user.id)
        return query

    # Update

    async def before_update(
        self, id: int, data: PlaceUpdateSchema
    ) -> PlaceUpdateContext:
        stmt = select(self.model.title, self.model.description).where(
            self.model.id == id
        )
        result = await self.session.execute(stmt)
        record = result.one_or_none()
        if record is None:
            raise self.not_found_error(id)

        title_changed = bool(data.title and data.title != record.title)
        description_changed = bool(
            data.description and data.description != record.description
        )
        return PlaceUpdateContext(
            trigger_embedding=description_changed or title_changed
        )

    async def after_update(
        self, id: int, data: PlaceUpdateSchema, context: PlaceUpdateContext
    ) -> None:
        if context.trigger_embedding:
            place_embedding(id)

    async def auth_put(
        self, user: UserReadSchema, id: int | str, form: PlacePutSchema
    ) -> None:
        if user.isAdmin:
            return

        id = self.parse_id(id)
        where: WhereFilters[PlaceSearchableFields] = {
            "id": self.eq(id),
            "creatorId": self.eq(user.id),
        }
        exists = await self.exists(where)
        if not exists:
            raise ApiError(
                HTTPStatus.UNAUTHORIZED,
                "Access denied",
                dict(message=f"Cannot access place {id}"),
            )

    async def update_embedding(self, id: int, embedding: list[float]) -> None:
        stmt = (
            update(self.model)
            .where(self.model.id == id)
            .values(embedding=text(f"'{json.dumps(embedding)}'::vector"))
        )
        await self.session.execute(stmt)
        await self.session.commit()

    async def embed(self, id: int) -> list[float]:
        stmt = select(self.model.title, self.model.description).where(
            self.model.id == id
        )
        result = await self.session.execute(stmt)
        row = result.one_or_none()

        if not row:
            raise ApiError(
                HTTPStatus.NOT_FOUND,
                f"No place with id {id} found in the database",
            )

        text = f"{row.title} - {row.description}"
        embedding = await hf_client.embed_text(text)

        try:
            await self.update_embedding(id, embedding)
        except Exception as err:
            raise ApiError(
                HTTPStatus.INTERNAL_SERVER_ERROR,
                "embedding failed",
                dict(place_id=id, message=str(err)),
            )

        return embedding

    # Delete

    async def before_delete(self, id: int) -> PlaceDeleteContext:
        stmt = select(self.model.id, self.model.image_url).where(
            self.model.id == id
        )
        result = await self.session.execute(stmt)
        record = result.one_or_none()
        if record is None:
            raise self.not_found_error(id)

        return PlaceDeleteContext(image_url=record.image_url)

    async def after_delete(self, id: int, context: PlaceDeleteContext) -> None:
        if context.image_url:
            cloud_storage.delete_file(context.image_url)

    async def auth_delete(self, user: UserReadSchema, id: int | str) -> None:
        if user.isAdmin:
            return

        id = self.parse_id(id)
        where: WhereFilters[PlaceSearchableFields] = {
            "id": self.eq(id),
            "creatorId": self.eq(user.id),
        }
        exists = await self.exists(where)
        if not exists:
            raise ApiError(
                HTTPStatus.UNAUTHORIZED,
                "Access denied",
                dict(message=f"cannot access place {id}"),
            )
